### DISCUSSION TITLE
Buildroot Toolchain building

### DISCUSSION THREAD

**Discussion by loredous** [mdubinko/gsoc-contrib#14] [Ideas]
The current 1.3 image for the GSOC seems to be built using an older version of BuildRoot (buildroot-2020.02-rc3), using an older kernel and I believe being based on the Cubieboard2 defconfig values. One of the items that will be necessary to allow better 3rd party development for the GSOC is going to be a working Buildroot configuration and toolchain. In pursuit of that, I have been working to determine what configuration would be needed there.

My current findings:
-  Using the cubieboard2 defconfig values does seem to build a toolchain and binaries that function properly on the GSOC (make cubieboard2_defconfig) ref:http://docs.cubieboard.org/tutorials/cb1/installation/cb1_making_embedded_linux_with_buildroot
-  The SDcard images created by Buildroot in the cubieboard2 configuration are of the same structure as the official images, having the same 3 partitions (boot/root/compressed rootfs image) and being bootable on the GSOC
-  If you overwrite the existing rootfs partition with a buildroot built ext4 image, you can replace the running rootfs on an SDcard boot
- If you replace the compressed rootfs on partition 3, you can get it to write to the internal GSOC storage

The major issue I have currently been working on is finding the proper uBoot and kernel configuration settings to be able to replace these portions of the official image. The basic problem comes down to the defconfig settings being for a cubieboard2, which does not have all the extra bits and bobs of the Einstein board (emmc storage, etc) and thus none of that is configured in the buildroot defconfig. It sounds like  @RAndrewThomas has some experience with Buildroot images. If anyone can come up with a working set of configurations, it would push the project forward significantly.

**Comment by n5rgn** [comment]
A hexdump of "boot.scr" from the gsoc's update image should get you the text of the gsoc's boot.cmd.  Quite different from the cubieboard2.  Copied here:

```
setenv bootargs console=ttyS0,115200 root=/dev/mmcblk0p2 rootwait panic=10 drm_kms_helper.drm_fbdev_overalloc=300
load mmc 0:1 0x43000000 ${fdtfile} || load mmc 0:1 0x43000000 boot/${fdtfile}
load mmc 0:1 0x42000000 uImage || load mmc 0:1 0x42000000 boot/uImage
bootm 0x42000000 - 0x43000000
```


On my end, I copied the buildroot cubieboard2 config files and whatnot to new "einsteina20" instances and am working from there.

I'm working on a buildroot config.  Started with the cubieboard2_defconfig with edits - not much there as you've seen.  The next step has been to examine the files in the extracted root.tar from the update image to determine what libraries, binaries, etc they built in and select options in buildroot "make menuconfig" that I think will fit.  Tedious.

I don't what your availability to work on this is, but after 2 years of work slowdown and a bunch of furlough this year, suddenly I get this new toy (GSOC) and my employer gets some big orders.  I'll continue to work on this as I can, but my schedule has tightened up a bunch.

**Reply by n5rgn** [reply]
Ugh.  I've found some source purporting to be for the einsteinA20 and CubieAIO-A20.  Trying to verify if its useful or not.

**Reply by loredous** [reply]
It sounds like you are definitely on the right path here, and it also sounds like you have more experience than me in this kind of thing, which is good. I'm doing a similar investigation, with looking at what is in the existing rootfs environment and seeing what needs to be included in the new one at a minimum. 

As for my availability, it changes from week to week depending on the work I'm doing at my day job, but I typically try to find some time while playing "hurry up and wait" to look into the GSOC stuff.

**Reply by n5rgn** [reply]
The lichee Einstein-A20 buildroot/source tree I found appears to be Cubie official, but as a tar gzip of a git repo they hosted on mega.nz.  Heck of a way to distribute code.  I can point you to it but I'm still trying to determine how relevant it is.

In the meantime I'm cross-checking various A20 and Cubie sources with what's in the GSOC image to ID drivers used and build a device tree definition.  Similar with U-Boot.  Between the Cubie sources and allwinner stuff in general, I think we have / can get most everything we need, the trick being what's up-to-date and relevant.  Also sundries off of the Einstein board like the Cirrus "sound card" (sources found) and the gsoc pushbuttons (no known sources, may have to ignore that one for a while).

**Comment by NeonFux** [comment]
https://github.com/d3cker/gsoc-puhumod
Debian on GSOC
